ouroboros-network/src/Ouroboros/Network/Mux/Types.hs

```diff
@@ -10,31 +10,78 @@ module Ouroboros.Network.Mux.Types (
     , ProtocolEnum (..)
     , MiniProtocolId (..)
     , MiniProtocolMode (..)
+    , MuxBearer (..)
+    , MuxBearerState (..)
     , MuxError (..)
     , MuxErrorType (..)
+    , MuxStyle (..)
     , MuxSDU (..)
     , MuxSDUHeader (..)
+    , NetworkMagic (..)
     , PerMuxSharedState (..)
     , RemoteClockModel (..)
     , TranslocationServiceRequest (..)
+    , Version (..)
+    , versionToVersionNumber
+    , versionMagic
+    , VersionNumber (..)
     , Wanton (..)
     ) where
 
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import           Codec.Serialise.Class
 import           Control.Exception
+import           Data.Array
 import qualified Data.ByteString.Lazy as BL
+import           Data.Ix (Ix (..))
 import           Data.Word
-import           Data.Ix (Ix(..))
-import           Data.Array
 import           GHC.Stack
 import           Text.Printf
 
-import           Control.Exception (throw, ArrayException(IndexOutOfBounds))
 import           Control.Monad.Class.MonadSTM
 import           Control.Monad.Class.MonadTimer
 import           Ouroboros.Network.Channel
 
 newtype RemoteClockModel = RemoteClockModel { unRemoteClockModel :: Word32 } deriving Eq
 
+data VersionNumber = VersionNumber0
+                   | VersionNumber1
+                   deriving (Eq, Ord, Show)
+
+newtype NetworkMagic = NetworkMagic { unNetworkMagic :: Word32 } deriving (Eq, Show, Ord)
+
+instance Serialise NetworkMagic where
+    encode (NetworkMagic nm) = encode nm
+    decode = NetworkMagic <$> decode
+
+data Version = Version0 !NetworkMagic -- Plain NodeToNode
+             | Version1 !NetworkMagic -- NodeToClient
+             deriving (Eq, Ord)
+
+versionMagic :: Version -> NetworkMagic
+versionMagic (Version0 nm) = nm
+versionMagic (Version1 nm) = nm
+
+instance Show Version where
+    show (Version0 m) = printf "Version 0 NetworkMagic %d" $ unNetworkMagic m
+    show (Version1 m) = printf "Version 1 NetworkMagic %d" $ unNetworkMagic m
+
+instance Serialise Version where
+    encode (Version0 m) = CBOR.encodeWord 0 <> encode m
+    encode (Version1 m) = CBOR.encodeWord 1 <> encode m
+
+    decode = do
+        vn <- CBOR.decodeWord
+        case vn of
+             0 -> Version0 . NetworkMagic <$> decode
+             1 -> Version1 . NetworkMagic <$> decode
+             n -> throw $ MuxError MuxControlUnknownVersion ("unknown version " ++ show n) callStack
+
+versionToVersionNumber :: Version -> VersionNumber
```

Comment:
I am not convienced that we really need this. What `Map VersionNumber Version` buys us vs `[Version]`? I guess that the version map is bound (no duplicate records).  Maybe we could express this nicely in `cborg` decoder of `[Version]`

Reply:
It is used to ensure that there are no duplicate Versions provided.